### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by scifi6546**
I am refactoring the code base to allow the IO library to be generalized outside of a gameboy emulator. That is not done yet but I plan on being able to quickly write an emulator for a system and the emulator would be able to deploy to native and webgl.

### CODE REVIEW

File: gameboy_opengl/src/lib.rs [hunk 1]
```diff
@@ -51,12 +51,11 @@ pub fn start(rom: Vec<u8>) -> Result<(), String> {
...
 
-    let mut cartridge = Cartridge::from_rom(rom);
-    load_ram_save_data(&mut cartridge).map_err(|e| format!("{:?}", e))?;
-    load_timestamp_data(&mut cartridge).map_err(|e| format!("{:?}", e))?;
-
     let rtc = Box::new(NativeRTC::new());
-    let mut emulator = Emulator::from_cartridge(cartridge, rtc);
+    let mut emulator = Gameboy::from_rom(rom, rtc);
+
+    load_ram_save_data(emulator.get_cartridge_mut());
```
Comment (benkonz) [inline gameboy_opengl/src/lib.rs]: We don't want to loose error information by ignoring the `Result` returned by the `load` methods. Let's add back the `.map_err(|e| format!("{:?}", e))?;` to these function calls.

File: src/lib.rs [hunk 2]
```diff
@@ -1,4 +1,8 @@
 #[cfg(target_arch = "wasm32")]
+extern crate gameboy_opengl_web;
```
Comment (benkonz) [inline src/lib.rs]: Rust 2018 edition makes these `extern crate` optional (unless you need it for `#[macro_use]`), so we can delete

```
#[cfg(target_arch = "wasm32")]
extern crate gameboy_opengl_web;
```

File: gameboy_opengl/src/lib.rs (hunk 1 above)
Comment (scifi6546) [reply gameboy_opengl/src/lib.rs]: I added in the map_error methods to the load functions. Must have missed the change when resolving merge conflicts.

File: src/lib.rs (hunk 2 above)
Comment (scifi6546) [reply src/lib.rs]: Deleted the Extern crate.

File: gameboy_opengl_web/src/lib.rs [hunk 3]
```diff
@@ -25,8 +25,9 @@ use stdweb::web::{document, window, Element, IEventTarget, TypedArray};
 use stdweb::Value;
 
 struct EmulatorState {
-    emulator: Emulator,
-    controller: Controller,
+    gameboy: Gameboy,
+    //emulator: Emulator,
```
Comment (benkonz) [inline gameboy_opengl_web/src/lib.rs]: We can just remove these two lines

File: src/lib.rs [hunk 4]
```diff
@@ -1,4 +1,6 @@
 #[cfg(target_arch = "wasm32")]
+pub use gameboy_opengl_web::DOMInfo;
```
Comment (benkonz) [inline src/lib.rs]: This import isn't needed

File: gameboy_opengl/src/lib.rs [hunk 5]
```diff
@@ -4,7 +4,7 @@ mod screen;
 use crate::native_rtc::NativeRTC;
```
Comment (benkonz) [inline gameboy_opengl/src/lib.rs]: This file needs to be updated to use the new controller functions in the `Gameboy` struct. The way it is now, the controls don't work on the native version

**Comment by scifi6546** [comment]
Yes I want to be able to generalize the code base to other consoles. Eventually I would like to emulate a C64 or some other home micro computer.

File: gameboy_opengl/src/lib.rs (hunk 5 above)
Comment (scifi6546) [reply gameboy_opengl/src/lib.rs]: It looks like the controls are now plugged back into the gameboy struct. There is an existing bug with the up and down arrow buttons not working that I should fix.

Comment (benkonz) [reply gameboy_opengl/src/lib.rs]: The controls seem to work fine for me. What game are you testing with?

Comment (scifi6546) [reply gameboy_opengl/src/lib.rs]: I am testing with a pokemon rom. The rom might be corrupted

Comment (scifi6546) [reply gameboy_opengl/src/lib.rs]: Tested it with a more official ROM and it is working fine

**Pull request merged**